File: 23. Permutation in String.cpp
// Topic: Permutation in String (check if s2 contains permutation of s1)
#include <bits/stdc++.h>
using namespace std;

bool checkInclusion(string s1, string s2)
{
    if (s1.size() > s2.size())
        return false;
    vector<int> cnt1(26, 0), cnt2(26, 0);
    for (int i = 0; i < s1.size(); ++i)
    {
        cnt1[s1[i] - 'a']++;
        cnt2[s2[i] - 'a']++;
    }
    if (cnt1 == cnt2)
        return true;
    for (int i = s1.size(); i < s2.size(); ++i)
    {
        cnt2[s2[i] - 'a']++;
        cnt2[s2[i - s1.size()] - 'a']--;
        if (cnt1 == cnt2)
            return true;
    }
    return false;
}

int main()
{
    string s1 = "ab";
    string s2 = "eidbaooo";
    cout << (checkInclusion(s1, s2) ? "Yes" : "No") << '\n';
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1)

File: 23. Permutation in String.py
# Topic: Permutation in String (check if s2 contains permutation of s1)

from collections import Counter

def check_inclusion(s1, s2):
    if len(s1) > len(s2): return False
    cnt1 = [0]*26
    cnt2 = [0]*26
    for i in range(len(s1)):
        cnt1[ord(s1[i]) - ord('a')] += 1
        cnt2[ord(s2[i]) - ord('a')] += 1
    if cnt1 == cnt2: return True
    for i in range(len(s1), len(s2)):
        cnt2[ord(s2[i]) - ord('a')] += 1
        cnt2[ord(s2[i-len(s1)]) - ord('a')] -= 1
        if cnt1 == cnt2: return True
    return False

if __name__ == '__main__':
    s1 = 'ab'
    s2 = 'eidbaooo'
    print('Yes' if check_inclusion(s1,s2) else 'No')

# Time Complexity: O(n)
# Space Complexity: O(1)
